exec/trace.go

```diff
@@ -130,12 +141,36 @@ func (t *tracer) Event(mach *sliceMachine, subject interface{}, ph string, args
 		event.Name = name.String()
 		event.Cat = "invocation"
 		key := compileKey{mach.Addr, arg.Index}
+		t.assignTid(mach, ph, t.compileEvents[key], &event)
 		t.compileEvents[key] = append(t.compileEvents[key], event)
 	default:
 		panic(fmt.Sprintf("unsupported subject type %T", subject))
 	}
 }
 
+// assignTid assigns a thread ID to event, using mach's tid pool and type of
+// event. events is the slices of existing relevant events, e.g.
+// t.taskEvents[arg].
+func (t *tracer) assignTid(mach *sliceMachine, ph string, events []traceEvent, event *traceEvent) {
+	event.Tid = 0
+	tidPool := t.machineTidPools[mach]
+	switch ph {
+	case "B":
+		event.Tid = tidPool.Acquire()
+		t.machineTidPools[mach] = tidPool
+	case "E":
+		if len(events) == 0 {
+			break
+		}
+		lastEvent := events[len(events)-1]
+		if lastEvent.Ph != "B" {
+			break
+		}
+		event.Tid = lastEvent.Tid
+		tidPool.Release(event.Tid)
```

Comment:
If I'm understanding correctly, this will cause panics on event streams with unpaired Es, while existing code (in `appendCoalesce`) acceptingly dropped them. This change in behavior isn't too concerning, as long as it's not actually happening anywhere yet, though I'm not sure if there are tests for event stream correctness, etc.

It does make me wonder, would it make sense to move thread id assignment into coalescing? I think lookback from E to the last B is a coalesce-y thing, so it'd fit well there. I don't think it'd make things more concise (you'd still need to keep this per-machine state), but it might be clearer.

Reply:
Can you describe the panic case you see? I assume you are talking about panicking in `Release`. I see two possibilities for an unpaired "E" -- we see an "E":

1. `events` is empty: we'll break before the `Release`.
2. `events` is non-empty and the previous event is non-"B": we'll break before the `Release`.

What am I missing?

Reply:
As for moving thread ID assignment into coalescing, I think that would complicate coalescing since we'd have to traverse events in time order, across event types(?). I actually originally had thread IDs passed in, so that callers of `Event` would control thread IDs. I actually think this makes the most ultimate sense, as callers would define the threads of work. I ended up isolating it here because it was encapsulated the thread ID concern, figuring that if we needed it in the future, we could push it out easily.

Maybe I'm missing a more simple approach.

Reply:
Ah, I think you're right about the lack of panicking.

Ok, it sounds like moving this to coalescing isn't really going to simplify things, so I think it's good as-is.